### PR TITLE
feat(addon-preview): add new package

### PR DESCRIPTION
**Pull request by MarsiBarsi**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [ ] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/)
-   [ ] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

-   [ ] Bugfix
-   [ ] Feature
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Other... Please describe:

## What is the current behavior?

Closes #

## Does this PR introduce a breaking change?

-   [ ] Yes
-   [ ] No

### CODE REVIEW

File: projects/addon-preview/components/preview/preview.style.less [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    width: 100%;
+    height: 100%;
+    user-select: none;
+}
+
+.header {
+    position: absolute;
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.style.less:14]: needs to be fixed

File: projects/demo-integrations/cypress/tests/addon-preview/addon-preview.spec.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+describe('Addon preview', () => {
+    beforeEach(() => {
+        cy.viewport(1400, 720);
+        cy.goToDemoPage('components/preview');
```
[the remark this answers is not in the document]
Comment (vladimirpotekhin) [reply projects/demo-integrations/cypress/tests/addon-preview/addon-preview.spec.ts:4]: thanks, fixed

File: projects/addon-preview/components/package.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+    "ngPackage": {
+        "lib": {
+            "entryFile": "index.ts",
+            "styleIncludePaths": [
+                "../../../core/styles"
```
Comment (nsbarsukov) [inline projects/addon-preview/components/package.json:6]: i guess it should be:

```suggestion
                "../../core/styles"
```

File: projects/addon-preview/components/preview-dialog/preview-dialog.component.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import {ChangeDetectionStrategy, Component, Inject} from '@angular/core';
+import {TuiDialog} from '@taiga-ui/cdk';
+import {tuiSlideInTop} from '@taiga-ui/core';
+import {POLYMORPHEUS_CONTEXT} from '@tinkoff/ng-polymorpheus';
+import {PreviewDialogOptions} from './preview-dialog-options';
+
+@Component({
+    selector: 'preview-dialog',
```
Comment (nsbarsukov) [inline projects/addon-preview/components/preview-dialog/preview-dialog.component.ts:8]:
```suggestion
    selector: 'tui-preview-dialog',
```

File: projects/addon-preview/components/preview/action/preview-action.style.less [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    margin-left: 0.3125rem;
+    margin-right: 0.3125rem;
+}
+
+.t-button {
+    .preview-item();
+    .transition(background);
+
+    &:hover {
+        background: rgba(159, 159, 159, 0.86);
```
Comment (nsbarsukov) [inline projects/addon-preview/components/preview/action/preview-action.style.less:15]: do we have similar color from our palette ?

File: projects/addon-preview/components/preview/action/preview-action.style.less [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    :host:not([data-size='fullscreen']) & {
+        animation: tuiFadeIn 0.3s;
```
Comment (nsbarsukov) [inline projects/addon-preview/components/preview/action/preview-action.style.less:19]:
```suggestion
        animation: tuiFadeIn var(--tui-duration);
```

File: projects/addon-preview/components/preview/mixin.style.less [hunk 7]
```diff
@@ -0,0 +1,5 @@
+.preview-item() {
+    border-radius: 1rem;
+    background: rgba(104, 104, 104, 0.96);
```
Comment (nsbarsukov) [inline projects/addon-preview/components/preview/mixin.style.less:3]: do we have similar color from our palette ?

File: projects/addon-preview/components/preview/preview.template.html [hunk 8]
```diff
@@ -0,0 +1,52 @@
+<ng-container *ngIf="texts$ | async as texts">
+    <section
+        #contentWrapper
+        class="wrapper"
```
Comment (nsbarsukov) [inline projects/addon-preview/components/preview/preview.template.html:4]: `t-` prefix

File: projects/demo/src/modules/components/preview/preview.template.html [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                Add <code>TuiDialogModule</code> to main module:
+
+                <tui-doc-code
+                    filename="app.module.ts"
+                    [code]="exampleImportHostModule"
+                ></tui-doc-code>
+            </li>
+
+            <li>
+                Import <code>TuiPreviewModule</code> to the module, witch will
```
Comment (nsbarsukov) [inline projects/demo/src/modules/components/preview/preview.template.html:59]:
```suggestion
                Import <code>TuiPreviewModule</code> to the module, which will
```

**Review by nsbarsukov — CHANGES_REQUESTED** [review]
🔥

File: projects/addon-preview/components/package.json (hunk 3 above)
Comment (nsbarsukov) [reply projects/addon-preview/components/package.json:6]: but check it, please)

Comment (vladimirpotekhin) [reply projects/addon-preview/components/package.json:6]: no, with 2 '/' build falls)

File: projects/addon-preview/components/preview/mixin.style.less (hunk 7 above)
Comment (vladimirpotekhin) [reply projects/addon-preview/components/preview/mixin.style.less:3]: No, unfortunately
We use the same colors in the dialog, but not from our palette

File: projects/addon-preview/components/preview/action/preview-action.style.less (hunk 5 above)
Comment (vladimirpotekhin) [reply projects/addon-preview/components/preview/action/preview-action.style.less:15]: see [comment](https://github.com/TinkoffCreditSystems/taiga-ui/pull/350#discussion_r741807150)

File: projects/demo/src/modules/components/preview/examples/1/index.less [hunk 10]
```diff
@@ -0,0 +1,19 @@
+@import 'taiga-ui-local';
+
+.content {
+    .text-body-xl();
+    background-color: rgb(242, 242, 242);
```
Comment (nsbarsukov) [inline projects/demo/src/modules/components/preview/examples/1/index.less:5]:
```suggestion
    background-color: var(--tui-base-01);
```

to improve dark mode:

example1

**Review by nsbarsukov — APPROVED** [review]

File: projects/demo/src/modules/components/preview/examples/1/index.less (hunk 10 above)
Comment (vladimirpotekhin) [reply projects/demo/src/modules/components/preview/examples/1/index.less:5]: thanks!

File: projects/addon-preview/components/preview/preview.component.ts [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        }),
+    );
+
+    readonly wrapperTransform$ = combineLatest([
+        this.coordinates$.pipe(map(([x, y]) => `${x}px, ${y}px`)),
+        this.zoom$,
+        this.rotation$,
+    ]).pipe(
+        map(([translate, zoom, rotation]) =>
+            this.sanitizer.bypassSecurityTrustStyle(
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:89]: Please check if this is required, maybe it was only needed before Angular 9

File: projects/addon-preview/components/preview/preview.component.ts [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        readonly texts$: Observable<LanguagePreview['previewTexts']>,
+    ) {}
+
+    rotate() {
+        this.rotation$.next(this.rotation$.value - ROTATION_ANGLE);
+    }
+
+    onPan(delta: [number, number]) {
+        this.coordinates$.next(
+            this.getGuardedCoordinates(
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:114]: Do you think maybe `scan` is a better solution?

File: projects/addon-preview/components/preview/preview.component.ts [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        boxWidth: number,
+    ): number {
+        const bigSize =
+            contentHeight > boxHeight * INITIAL_SCALE_COEF ||
+            contentWidth > boxWidth * INITIAL_SCALE_COEF;
+
+        return bigSize
+            ? round(
+                  Math.min(
+                      (this.elementRef.nativeElement.clientHeight * INITIAL_SCALE_COEF) /
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:168]: let's destruct into clientHeight and clientWidth into const prior to that

File: projects/addon-preview/components/preview/preview.component.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+        const oldScale = this.zoom$.value;
+        const newScale = clamp(this.zoom$.value + delta, this.minZoom, 2);
+
+        const center = this.getScaleCenter(
+            {clientX, clientY},
+            this.coordinates$.value,
+            this.zoom$.value,
+        );
+
+        this.zoom$.next(newScale);
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:224]: Let's move it down to the other `next` so the resulting action is all in the same place

File: projects/addon-preview/components/preview/preview.component.ts [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+                          contentHeight,
+                      (this.elementRef.nativeElement.clientWidth * INITIAL_SCALE_COEF) /
+                          contentWidth,
+                  ),
+                  2,
+              )
+            : 1;
+    }
+
+    private initClickSubscription(element: HTMLElement) {
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:178]: I don't like that we do this. If there's no way to make it cleaner I say let's just drop this feature. Sergey Novoselov doesn't mind.

File: projects/addon-preview/components/preview/preview.component.ts (hunk 14 above)
Comment (vladimirpotekhin) [reply projects/addon-preview/components/preview/preview.component.ts:182]: We have to update zoom first and then compute new coordinates, so the order matters. ( Maybe split this method in two?

File: projects/addon-preview/components/preview/loading/preview-loading.component.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import {ChangeDetectionStrategy, Component} from '@angular/core';
+
+@Component({
+    selector: 'tui-preview-loading',
+    templateUrl: './preview-loading.template.html',
+    styleUrls: ['./preview-loading.style.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TuiPreviewLoadingComponent {}
```
[the remark this answers is not in the document]
Comment (vladimirpotekhin) [reply projects/addon-preview/components/preview/loading/preview-loading.component.ts]: I removed it 👍

File: projects/demo/src/modules/components/preview/examples/import/template.txt [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    <tui-preview>
+        <tui-preview-title>{{title}}</tui-preview-title>
+        <tui-preview-pagination
+            [lastIndex]="1"
+            [(index)]="index"
+        ></tui-preview-pagination>
+
+        <tui-preview-action
+            icon="tuiIconTrash"
+            title="Удалить"
```
Comment (splincode) [inline projects/demo/src/modules/components/preview/examples/import/template.txt:11]: i18n?

File: projects/demo/src/modules/components/preview/examples/import/template.txt [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+            (click)="delete()"
+        ></tui-preview-action>
+        <tui-preview-action
+            icon="tuiIconDownload"
+            title="Скачать"
```
Comment (splincode) [inline projects/demo/src/modules/components/preview/examples/import/template.txt:16]: i18n?

File: projects/addon-preview/components/preview/pagination/preview-pagination.component.ts [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+@Component({
+    selector: 'tui-preview-pagination',
+    templateUrl: './preview-pagination.template.html',
+    styleUrls: ['./preview-pagination.style.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TuiPreviewPaginationComponent {
+    @Input()
+    @tuiDefaultProp()
+    lastIndex = 1;
```
Comment (waterplea) [inline projects/addon-preview/components/preview/pagination/preview-pagination.component.ts:24]: Pagination has `length`, let's rename for consistency.

File: projects/addon-preview/components/preview/pagination/preview-pagination.style.less [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+    .preview-item();
+    .text-body-s();
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+    width: 6.25rem;
+}
+
+.t-arrow {
+    .transition(background);
```
Comment (waterplea) [inline projects/addon-preview/components/preview/pagination/preview-pagination.style.less:14]: Can we reuse tuiPreviewAction here as well?

File: projects/addon-preview/components/preview/preview.component.ts [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+
+    readonly zoom$ = new BehaviorSubject<number>(this.minZoom);
+    readonly rotation$ = new BehaviorSubject<number>(0);
+    readonly coordinates$ = new BehaviorSubject<readonly [number, number]>(
+        EMPTY_COORDINATES,
+    );
+
+    readonly transitioned$ = merge(
+        dragAndDropFrom(this.elementRef.nativeElement).pipe(
+            map(state => state.stage !== TuiDragStage.Continues),
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:58]: nit: not hip

File: projects/addon-preview/components/preview/preview.component.ts [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+        typedFromEvent(this.elementRef.nativeElement, 'touchmove', {passive: true}).pipe(
+            mapTo(false),
+        ),
+        typedFromEvent(this.elementRef.nativeElement, 'wheel', {passive: true}).pipe(
+            mapTo(false),
+        ),
+    );
+
+    readonly cursor$ = dragAndDropFrom(this.elementRef.nativeElement)
+        .pipe(startWith(null))
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.component.ts:69]: Let's move it down and write `startWith('initial')` then we can make `map` hip

File: projects/addon-preview/components/preview/preview.template.html [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        <div class="t-actions">
+            <ng-content select="[tuiPreviewAction]"></ng-content>
+        </div>
+    </header>
+
+    <footer class="t-footer">
+        <button
+            *ngIf="rotatable"
+            tuiIconButton
+            tuiPreviewAction
```
Comment (waterplea) [inline projects/addon-preview/components/preview/preview.template.html:37]: Doesn't margin move it off center?

File: projects/addon-preview/components/preview/zoom/preview-zoom.template.html [hunk 24]
```diff
@@ -0,0 +1,56 @@
+<ng-container *ngIf="zoomTexts$ | async as texts">
+    <section class="t-zoom">
+        <button
+            tuiIconButton
+            size="s"
+            appearance=""
```
Comment (waterplea) [inline projects/addon-preview/components/preview/zoom/preview-zoom.template.html:6]: Can we reuse tuiPreviewAction here?

**Review by splincode — APPROVED** [review]

File: projects/addon-preview/components/preview/preview.template.html (hunk 23 above)
Comment (vladimirpotekhin) [reply projects/addon-preview/components/preview/preview.template.html:37]: margin removed from tuiPreviewAction 👍

**Review by waterplea — APPROVED** [review]

**Pull request merged**